File: Permutations.cpp
class Solution {
public:
    /**
     * @param nums: A list of integers.
     * @return: A list of permutations.
     */
    vector<vector<int> > permute(vector<int> nums) {
        // write your code here
        vector<vector<int>> ret;
        if (nums.empty())
        {
            return ret;
        }
        sort(nums.begin(), nums.end());
        do
        {
            ret.emplace_back(nums);
        } while (nextPermutation(nums));
        return ret;
    }
    bool nextPermutation(vector<int> &nums)
    {
        int len = nums.size();
        int i = len - 1;
        while (i > 0 && nums[i] < nums[i - 1])
        {
            --i;
        }
        
        if (i > 0)
        {
            int j = len - 1;
            while (nums[j] < nums[i - 1])
            {
                --j;
            }
            swap(nums[i - 1], nums[j]);
            reverse(nums.begin() + i, nums.end());
            return true;
        }
        else
        {
            return false;
        }
        
    }
};

File: Permutations.py
class Solution:
    """
    @param nums: A list of Integers.
    @return: A list of permutations.
    """
    def permute(self, nums):
        # write your code here
        def helper(left, visited, tmp, ret, length):
            if left == 0:
                ret.append(tmp[::])
            else:
                for i in range(0, length):
                    if not visited[i]:
                        tmp.append(nums[i])
                        visited[i] = True
                        helper(left - 1, visited, tmp, ret, length)
                        visited[i] = False
                        tmp.pop()
        ##error
        if not nums:
            return []
        nums.sort()
        length = len(nums)
        visited = [False] * length
        tmp, ret = [], []
        helper(length, visited, tmp, ret, length)
        return ret
